path.go

```diff
@@ -1,9 +1,313 @@
+		return err
+	}
+
+	if err := binary.Write(w, binary.BigEndian, hd.ForwardAmount); err != nil {
+		return err
+	}
+
+	if err := binary.Write(w, binary.BigEndian, hd.OutgoingCltv); err != nil {
+		return err
+	}
+
+	if _, err := w.Write(hd.ExtraBytes[:]); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Decodes populates the target HopData with the contents of a serialized
+// HopData packed into the passed io.Reader.
+func (hd *HopData) Decode(r io.Reader) error {
+	if _, err := io.ReadFull(r, hd.Realm[:]); err != nil {
+		return err
+	}
+
+	if _, err := io.ReadFull(r, hd.NextAddress[:]); err != nil {
+		return err
+	}
+
+	err := binary.Read(r, binary.BigEndian, &hd.ForwardAmount)
+	if err != nil {
+		return err
+	}
+
+	err = binary.Read(r, binary.BigEndian, &hd.OutgoingCltv)
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.ReadFull(r, hd.ExtraBytes[:]); err != nil {
+		return err
+	}
+
+	return err
+}
+
+// PayloadType denotes the type of the payload included in the onion packet.
+// Serialization of a raw HopPayload will depend on the payload type, as some
+// include a varint length prefix, while others just encode the raw payload.
+type PayloadType uint8
+
+const (
+	// PayloadLegacy is the legacy payload type. It includes a fixed 32
+	// bytes, 12 of which are padding, and uses a "zero length" (the old
+	// realm) prefix.
+	PayloadLegacy PayloadType = iota
+
+	// PayloadTLV is the new modern TLV based format. This payload includes
+	// a set of opaque bytes with a varint length prefix. The varint used
+	// is the same CompactInt as used in the Bitcoin protocol.
+	PayloadTLV
+)
+
+// HopPayload is a slice of bytes and associated payload-type that are destined
+// for a specific hop in the PaymentPath. The payload itself is treated as an
+// opaque data field by the onion router. The included Type field informs the
+// serialization/deserialziation of the raw payload.
+type HopPayload struct {
+	// Type is the type of the payload.
+	Type PayloadType
+
+	// Payload is the raw bytes of the per-hop payload for this hop.
+	// Depending on the realm, this pay be the regular legacy hop data, or
+	// a set of opaque blobs to be parsed by higher layers.
+	Payload []byte
+
+	// HMAC is an HMAC computed over the entire per-hop payload that also
+	// includes the higher-level (optional) associated data bytes.
+	HMAC [HMACSize]byte
+}
+
+// NewHopPayload creates a new hop payload given an optional set of forwarding
+// instructions for a hop, and a set of optional opaque extra onion bytes to
+// drop off at the target hop. If both values are not specified, then an error
+// is returned.
+func NewHopPayload(hopData *HopData, eob []byte) (HopPayload, error) {
+
+	var (
+		h HopPayload
+		b bytes.Buffer
+	)
+
+	// We can't proceed if neither the hop data or the EOB has been
+	// specified by the caller.
+	if hopData == nil && len(eob) == 0 {
+		return h, fmt.Errorf("either hop data or eob must " +
+			"be specified")
+	}
```

Comment:
should also fail if `hopData != nil && len(eob) > 0`?

Reply:
Yep! I wan to force the caller to only provide one. In retrospect, I think I favor an API where you provide one of the other, but it was done this way in order to minimize code impact in the router within `lnd`. Once this goes in, I may follow up with a partial re-design of the interface.